### PR TITLE
terraform-providers.bitbucket: reintroduce

### PR DESCRIPTION
**Pull request by arcnmx**
###### Motivation for this change

This provider was removed as part of #153015, but it's still useful!

I'm not sure if it counts as "blessed", but there's certainly [one obvious fork](https://github.com/DrFaust92/terraform-provider-bitbucket) that's being actively developed and maintained, so I've moved to it.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/cluster/terraform-providers/providers.json [hunk 1]
```diff
@@ -144,6 +144,15 @@
     "vendorSha256": null,
     "version": "1.12.2"
   },
+  "bitbucket": {
+    "owner": "DrFaust92",
+    "provider-source-address": "registry.terraform.io/nixpkgs/bitbucket",
```
Comment (zowoq) [inline pkgs/applications/networking/cluster/terraform-providers/providers.json:149]: This doesn't exist on the terraform registry?

**Review by zowoq — DISMISSED** [review]

**Comment by arcnmx** [comment]
> Can explain why you picked the `DrFaust92` provider please?
> 
> There are 15 `bitbucket` providers on the registry: https://registry.terraform.io/search/providers?q=bitbucket, looks like there are a couple of others that are more popular and have been updated in the last few months.

So, to start with forks of the original archived hashicorp provider...
- `aeirola` was the initial popular fork, but it's been deprecated/archived and `DrFaust92`'s appears to be an active continuation of its efforts to maintain the original provider
- `Runelab` is a copy of `aeirola` with unspecified changes (no commit history) and no maintenance efforts
- `a0s` also a copy of `aeirola`
- `dj-wasabi` is also a copy of `aeirola`
- `andsafe-AG` is also a copy of `aeirola`
- `Zensor-BE` is also a copy of `aeirola` with various changes released as a beta
- ` rogercastaneda ` is also a copy of `aeirola`...
- `cibiltechgroup` is a minor fork of the original hashicorp repo
- `vahubertcibiltech` is a copy of `cibiltechgroup` by a member of that org, kinda weird
- `a0s/bitbucket-drfaust92` is a fork containing changes currently pending PR merge upstream to `DrFaust92`

Basically, 90% of them are forks people made to submit a PR to `aeirola` and only published once, probably because they were impatient while waiting for PRs to merge. The only real contender was `aeirola` (now deprecated in favour of `DrFaust92`). Its popularity could be justified by the fact that 13/15 of the options are forks of either it or its predecessor, and it includes all of their changes!

Now that leaves the real alternatives, which both appear to be full rewrites:
- https://github.com/cgroschupp/terraform-provider-bitbucket
- https://github.com/zahiar/terraform-provider-bitbucket

The former was active for 2 months last year and abandoned since July, so I'm discounting it. The latter is probably worth considering! It is however an entirely new provider, so I decided to go with a fork of the hashicorp provider for compatibility/familiarity/stability/etc reasons.
Feel free to give your opinion on `zahiar/bitbucket` as an alternative.

**Comment by zowoq** [comment]
Thanks for detailed response.

https://github.com/gavinbunney/terraform-provider-bitbucketserver
Is this another alternative? Doesn't seem popular on the registry but it is the most starred on github. Does seem to be new rather than a fork so the same reasons also apply.

> Feel free to give your opinion on `zahiar/bitbucket` as an alternative.

I don't really have any opinion, mostly just want to make sure it isn't surprising to anyone and we have something to refer to if a future PR tries to switch another one because e.g. it is more popular on the registry.

This one seems fine but I'll wait for @zimbatm to have a look.

**Comment by arcnmx** [comment]
> Is this another alternative? Doesn't seem popular on the registry but it is the most starred on github. Does seem to be new rather than a fork so the same reasons also apply.

Oh sorry, yeah, I skimmed over those because it wasn't clear to me whether it was exclusive to a self-hosted bitbucket instance or not - but it seems like it likely doesn't support the public bitbucket cloud? It's also confusing given that [Atlassian has deprecated Bitbucket Server in favour of a more SaaS alternative](https://www.atlassian.com/migration/assess/journey-to-cloud) or something...

**Comment by arcnmx** [comment]
Yeah, I mean I can't guarantee that this particular person's fork will stand the test of time... but there's clearly a collective need for this provider as evidenced by the community that's built up around these forks and efforts to preserve the original archived repo - and as long as one does exist and remains functional, I certainly benefit from its inclusion in nixpkgs.

I'll keep an eye on the newer `zahiar` one that's popped up, as it very well could become a proper successor!

**Pull request merged**